### PR TITLE
Speed up storage iteration from within the runtime

### PR DESCRIPTION
**Pull request by koute**
Followup of https://github.com/paritytech/substrate/pull/13284 and https://github.com/paritytech/substrate/pull/13445

This PR makes iterating over storage from within the runtime **twice as fast**. (It cuts down the time required to ~52% according to the benchmark I've added.)

### How does it work?

The way storage iteration is done from within the runtime is through the use of the `Storage::next_key` host function. What this host function does is pretty simple: given a `key` it will return the very next key that follows it in lexicographic order. So call it in a loop, and there you go - you're iterating over the storage.

But this is somewhat wasteful since every time this host function is called it will traverse the storage trie from scratch, wasting a lot of time going through nodes it has already went through in the previous iteration. So the trick to speed it up is pretty simple: instantiate an iterator every time this host function is called, stash it somewhere, and then try to reuse it the next time the host function is called if the `key` that was passed matches the key which the iterator previously returned.

### CODE REVIEW

File: primitives/state-machine/src/trie_backend.rs [hunk 1]
```diff
@@ -184,15 +187,41 @@ where
...
+	fn default() -> Self {
+		Self { is_cached: false, last_key: Default::default(), iter: Default::default() }
+	}
+}
+
 /// Patricia trie-based backend. Transaction type is an overlay of changes to commit.
 pub struct TrieBackend<S: TrieBackendStorage<H>, H: Hasher, C = LocalTrieCache<H>> {
 	pub(crate) essence: TrieBackendEssence<S, H, C>,
+	#[cfg(feature = "std")]
+	next_storage_key_cache: Mutex<CachedIter<S, H, C>>,
```
Comment (cheme) [inline primitives/state-machine/src/trie_backend.rs:213]: would make sense to have it for no_std (behind refcell).
Why not `Mutex<Option<CachedIter<S, H, C>>`? (and rem `is_cached`).

**Review by cheme — APPROVED** [review]

File: primitives/state-machine/src/trie_backend.rs (hunk 1 above)
Comment (koute) [reply primitives/state-machine/src/trie_backend.rs]: Done.

**Comment by koute** [comment]
(Sorry for the force push; commits are exactly the same, I just accidentally didn't sign one of the intermediate commits and the only way to rectify that is with a force push.)

File: primitives/state-machine/src/trie_backend_essence.rs [hunk 2]
```diff
@@ -417,7 +417,8 @@ where
 
 	/// Return the next key in the trie i.e. the minimum key that is strictly superior to `key` in
 	/// lexicographic order.
```
Comment (bkchr) [inline primitives/state-machine/src/trie_backend_essence.rs:419]: Could you please update the docs?

**Review by bkchr — APPROVED** [review]
Nice! 

I also had this somewhere on my long list! Ty for doing this!

File: primitives/state-machine/src/trie_backend_essence.rs (hunk 2 above)
Comment (koute) [reply primitives/state-machine/src/trie_backend_essence.rs]: Not entirely sure if this is what you wanted, but done. (:

**Comment by koute** [comment]
bot merge

File: primitives/state-machine/src/trie_backend_essence.rs (hunk 2 above)
Comment (bkchr) [reply primitives/state-machine/src/trie_backend_essence.rs]: Yeah sorry, I should have been writing more here. Your update looks like what I wanted!

**Pull request merged**